File: src/java/magmac/app/compile/rule/NodeListRule.java
package magmac.app.compile.rule;

import magmac.app.compile.error.CompileResult;
import magmac.app.compile.error.CompileResultCollector;
import magmac.app.compile.node.MapNode;
import magmac.app.compile.node.Node;
import magmac.app.compile.node.NodeList;
import magmac.app.compile.node.NodeListCollector;
import magmac.app.compile.rule.divide.Divider;
import magmac.app.compile.rule.divide.FoldingDivider;
import magmac.app.compile.rule.fold.Folder;
import magmac.app.lang.ValueFolder;

public final class NodeListRule implements Rule {
    private final String key;
    private final Rule childRule;
    private final Divider divider;

    public NodeListRule(String key, Rule childRule, Divider divider) {
        this.key = key;
        this.childRule = childRule;
        this.divider = divider;
    }

    public static Rule Values(String key, Rule childRule) {
        return NodeListRule.createNodeListRule(key, new ValueFolder(), childRule);
    }

    public static Rule createNodeListRule(String key, Folder folder, Rule childRule) {
        return new NodeListRule(key, childRule, new FoldingDivider(folder));
    }

    @Override
    public CompileResult<Node> lex(String input) {
        return this.divider.divide(input)
                .map(this.childRule::lex)
                .collect(new CompileResultCollector<>(new NodeListCollector()))
                .mapValue((NodeList children) -> new MapNode().withNodeList(this.key, children));
    }

    @Override
    public CompileResult<String> generate(Node node) {
        return node.findNodeListOrError(this.key)
                .flatMapValue((NodeList list) -> list.join(this.divider.createDelimiter(), this.childRule::generate));
    }
}

File: src/web/magmac/app/compile/rule/NodeListRule.ts
import { CompileResult } from "../../../../magmac/app/compile/error/CompileResult";
import { CompileResultCollector } from "../../../../magmac/app/compile/error/CompileResultCollector";
import { MapNode } from "../../../../magmac/app/compile/node/MapNode";
import { Node } from "../../../../magmac/app/compile/node/Node";
import { NodeList } from "../../../../magmac/app/compile/node/NodeList";
import { NodeListCollector } from "../../../../magmac/app/compile/node/NodeListCollector";
import { Divider } from "../../../../magmac/app/compile/rule/divide/Divider";
import { FoldingDivider } from "../../../../magmac/app/compile/rule/divide/FoldingDivider";
import { Folder } from "../../../../magmac/app/compile/rule/fold/Folder";
import { ValueFolder } from "../../../../magmac/app/lang/ValueFolder";
export class NodeListRule {
	NodeListRule(key : String, childRule : Rule, divider : Divider) : public {break;break;break;;}
	Values(key : String, childRule : Rule) : Rule {return NodeListRule.createNodeListRule( key, new ValueFolder( ), childRule);;}
	createNodeListRule(key : String, folder : Folder, childRule : Rule) : Rule {return new NodeListRule( key, childRule, new FoldingDivider( folder));;}
	lex(input : String) : CompileResult<Node> {return this.divider.divide( input).map( this.childRule.lex).collect( new CompileResultCollector<>( new NodeListCollector( ))).mapValue( 0);;}
	generate(node : Node) : CompileResult<String> {return node.findNodeListOrError( this.key).flatMapValue( 0);;}
}
